src/AgentImplementation.sol

```diff
@@ -298,36 +293,29 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
     }
 
     function _chargeFeeByMsgValue() internal {
-        if (msg.value == 0) return;
-
-        // TODO: implement new msg.value fee charge mechanism
+        if (msg.value == 0) {
+            return;
+        } else {
+            uint256 feeRate = IRouter(router).feeRate();
+            address feeCollector = IRouter(router).feeCollector();
+            IParam.Fee memory fee = FeeLogic.getFee(_NATIVE, msg.value, feeRate, _NATIVE_FEE_META_DATA);
+            fee.charge(feeCollector);
```

Comment:
Does anyone feel the same that this is more like charging `feeCollector`?

Reply:
I agree, given that it's within `FeeLogic`, perhaps using `transfer` would be more straightforward since we're already in the context of handling fees. Or perhaps `chargeBy` would be more appropriate.

Reply:
I see. Will change the naming to `chargeBy`.